src/importer.cpp

```diff
@@ -0,0 +1,483 @@
+ *
+ * Resolves the full path to the given @p filename using the @p base.
+ *
+ * This function is only intended to work with local files.  It may not
+ * work with bases that use the 'file://' prefix.
+ *
+ * @param filename The @c std::string relative path from the base path.
+ * @param base The @c std::string location on local disk for determining the full path from.
+ *
+ * @return The full path from the @p base location to the @p filename
+ */
+std::string resolvePath(const std::string &filename, const std::string &base)
+{
+    // We can be naive here as we know what we are dealing with.
+    std::string path = base.substr(0, base.find_last_of('/') + 1) + filename;
+    return path;
+}
+
+bool checkUnitsForCycles(ModelPtr &model, std::vector<std::tuple<std::string, std::string, std::string>> &history)
+{
+    // Check through model for overlap with the current history so we can report cycles properly.
+    for (size_t u = 0; u < model->unitsCount(); ++u) {
+        auto units = model->units(u);
+        if (units->isImport()) {
+            ImportSourcePtr importSource = units->importSource();
+            auto h = std::make_tuple(units->name(), units->importReference(), importSource->url());
+            if (std::find(history.begin(), history.end(), h) != history.end()) {
+                history.emplace_back(h);
+                return false;
+            }
+        }
+    }
+    return true;
+}
+bool checkComponentForCycles(ModelPtr &model, std::vector<std::tuple<std::string, std::string, std::string>> &history)
+{
+    for (size_t c = 0; c < model->componentCount(); ++c) {
+        auto component = model->component(c);
+        if (component->isImport()) {
+            ImportSourcePtr importSource = component->importSource();
+            auto h = std::make_tuple(component->name(), component->importReference(), importSource->url());
+            if (std::find(history.begin(), history.end(), h) != history.end()) {
+                history.emplace_back(h);
+                return false;
+            }
+        }
+    }
+    return true;
+}
+
+bool Importer::ImporterImpl::doResolveImports(ModelPtr &model, const std::string &baseFile,
+                                              std::vector<std::tuple<std::string, std::string, std::string>> &history)
+{
+    for (size_t n = 0; n < model->unitsCount(); ++n) {
+        auto units = model->units(n);
+
+        if ((!resolveImport(units, units->name(), "units", baseFile, history)) && (!history.empty())) {
+            std::string msg = "Cyclic dependencies were found when attempting to resolve units in model '" + model->name() + "'. The dependency loop is:\n";
+            std::string spacer = "    ";
+            for (auto &h : history) {
+                msg += spacer + "units '" + std::get<0>(h) + "' imports '" + std::get<1>(h) + "' from '" + std::get<2>(h);
+                spacer = "',\n    ";
+            }
+            msg += "'.";
```

Comment:
Right now, this will result in something like:
```
Cyclic dependencies were found when attempting to resolve units in model 'circularImport1'. The dependency loop is:
    units 'i_am_cyclic' imports 'u2' from 'circularUnits_2.cellml',
    units 'u2' imports 'u3' from 'circularUnits_3.cellml',
    units 'u3' imports 'i_am_cyclic' from 'circularUnits_1.cellml',
    units 'i_am_cyclic' imports 'u2' from 'circularUnits_2.cellml'.
```

while I think it would look better as follows:
```
Cyclic dependencies were found when attempting to resolve units in model 'circularImport1'. The dependency loop is:
 - units 'i_am_cyclic' imports 'u2' from 'circularUnits_2.cellml';
 - units 'u2' imports 'u3' from 'circularUnits_3.cellml';
 - units 'u3' imports 'i_am_cyclic' from 'circularUnits_1.cellml'; and
 - units 'i_am_cyclic' imports 'u2' from 'circularUnits_2.cellml'.
```

Reply:
7b7ad654

Reply:
"and" is missing from the penultimate item. :)